### ISSUE TITLE
Remember for X days

### ISSUE THREAD

**Issue by WimVDK** [akeeba/loginguard#77]
Would it be possible in the next version to add a setting to remember for X days (on most sites i've encountered this is by default *30* days

**Comment by nikosdion** [comment]
I have already [considered and decided against](https://github.com/akeeba/loginguard/issues/28) this feature due to security concerns.

**Comment by WimVDK** [comment]
Don't fully understand the technical stuff mentioned there. Any chance you could "dumb it down" a bit. I'm not asking full blown details on the security concerns. Just some basic logic, as would try to understand why you choose against it while other sites implement it.

**Comment by nikosdion** [comment]
Google, Microsoft and other tech giants do offer a "remember me" feature, sort of. However, it's NOT based only on cookies. It's based on their knowledge of where you are and which devices you're using thanks to their operating systems, applications and services – not to mention in some cases the information they collect through third party sites. If they cannot determine this information with adequate accuracy, e.g. because you're using your browser with moderate to high privacy settings like I do, they ask you to validate your login every time.

I could easily implement a similar feature by avoiding to show LoginGuard's captive login page (the Two Step Verification page) when your login is resumed using cookie authentication i.e. when you're using the Remember Me feature. However, Joomla's Remember Me feature is based on cookies and is not tied to a specific browser or device. In case the attacker has exfiltrated your cookies (e.g. by means of your presence in a hostile network, malware etc) they can automatically log into your site impersonating you without having to provide a username, password or two step verification. That's incredibly bad for security.

This is the reason why Big Tech does not rely on a single data point (a pretty much anonymous cookie) to implement that feature. Their cookies are tied to a unique identifier of your browser and device as detected through their deep knowledge of what you're doing. When that assurance is disrupted – taking them to the equivalent situation of Joomla's Remember Me cookie – they _always_ ask you to validate your login. Therefore if we want to follow their very good security example we need to have LoginGuard always ask you for verification.

The technical information in the issue I linked you to was describing my preliminary thoughts about determining a risk factor for the login in the same way Big Tech does. The more I was thinking about it the more it became obvious that without a global surveillance network there is not enough secure information to make that decision which is why I decided to cancel development on that feature last year.

**Comment by WimVDK** [comment]
I see, so basically Joomla would first have to "update" it's cookie methods itself (as i believe it doesn't let you "remember" either, before you'd be fully able (without some massive workaround like you had to do with the captive login redirect ;), yeah i read that article and github note :p) to adequately implement it.

Guess it's the same reason why you do it this way over at the akeeba site.

Noted. Maybe in the future when joomla stops "adopting" your code :)

**Comment by nikosdion** [comment]
What I am saying is that it's impossible to "update" the cookie method. "Updating" in this context means that you need a global surveillance network which monitors the whereabouts of everyone in real time. Google knows where you are, who you are with and which devices you are using right now. Joomla cannot (and should not!) do that.

Regarding Joomla's privacy policy plugin they do use my code, partially. I was the first developer to demonstrate that creating a captive login page in Joomla is not only possible, it's also backwards compatible. I first described that in 2012 but I was told it wouldn't be the case – that's why Joomla has that awful kludge of a Two Factor Authentication code instead of Two Step Verification. I was supposed to work on that in 2013 but the restrictions placed upon core contributions made it impossible. So I gave up and started working on other things before I got back to that idea in 2016; that's how LoginGuard came to be.

**Comment by WimVDK** [comment]
Aha, understood! In that case you're right, we don't want joomla to "spy" on us the way can :)

Case closed in that case. Thanks for your time and expert explanation!
